File: Data Structures and Problem Solving Using Cpp 2nd/PrintInt.cpp

#ifdef USE_DOT_H
    #include <iostream.h>
#else
    #include <iostream>
    #if !defined( __BORLANDC__ ) || __BORLANDC__ >= 0x0530
        using namespace std;      // Borland has a bug
    #endif
#endif

#ifdef SAFE_STL
    #include "mystring.h"
    #include "StartConv.h"
#else
    #include <string>
    using std::string;
#endif

// Print n in base base, recursively.
// Precondition: n >= 0, 2 <= base <= MAX_BASE.

const string DIGIT_TABLE = "0123456789abcdef";
const int    MAX_BASE    = DIGIT_TABLE.length( );

void printIntRec( int n, int base )
{
    if( n >= base )
        printIntRec( n / base, base );
    cout << DIGIT_TABLE[ n % base ];
}

// Driver routine.
void printInt( int n, int base )
{
    if( base <= 1 || base > MAX_BASE )
        cerr << "Cannot print in base " << base << endl;
    else
    {
        if( n < 0 )
        {
            cout << "-";
            n = -n;
        }
        printIntRec( n, base );
    }
}

int main( )
{
    for( int i = 0; i <= 17; i++ )
    {
        printInt( -1000, i );
        cout << endl;
    }
    return 0;
}

#ifdef USE_STL
    #include "EndConv.h"
#endif

File: Data Structures and Problem Solving Using Java 4th/PrintInt.java
public final class PrintInt
{
    private static final String DIGIT_TABLE = "0123456789abcdef";
    private static final int    MAX_BASE    = DIGIT_TABLE.length( );

    // Print n in any base, recursively
	// Precondition: n >= 0, 2 <= base <= MAX_BASE
    private static void printIntRec( long n, int base )
    {
        if( n >= base )
            printIntRec( n / base, base );
        System.out.print( DIGIT_TABLE.charAt( (int) ( n % base ) ) );
    }

    // Driver routine
    public static void printInt( long n, int base )
    {
        if( base <= 1 || base > MAX_BASE )
            System.err.println( "Cannot print in base " + base );
        else
        {
            if( n < 0 )
            {
                n = -n;
                System.out.print( "-" );
            }
            printIntRec( n, base );
        }
    }

        // Simple test program
    public static void main( String [ ] args )
    {
        for( int i = 0; i <= 17; i++ )
        {
            printInt( 1000, i );
            System.out.println( );
        }
        printInt( 0x5DEECE66DL, 10 );
        System.out.println( );
    }
}
